### ISSUE TITLE
Errors in compiling pcl_ros

### ISSUE THREAD

**Issue by mfe7** [ethz-asl/voxblox-plusplus#55]
When running `catkin build mask_rcnn_ros depth_segmentation gsm_node`, building `pcl_ros` fails:
```
Failed     << pcl_ros:make                        [ Exited with code 2 ]                                                                                                                     
Failed    <<< pcl_ros                             [ 35.8 seconds ]                                                                                                                           
Abandoned <<< depth_segmentation                  [ Unrelated job failed ]                                                                                                                   
Abandoned <<< voxblox_ros                         [ Unrelated job failed ]                                                                                                                   
Abandoned <<< gsm_node                            [ Unrelated job failed ]                                                                                                                   
[build] Summary: 20 of 24 packages succeeded.                                                                                                                                                
[build]   Ignored:   13 packages were skipped or are blacklisted.                                                                                                                            
[build]   Warnings:  9 packages succeeded with warnings.                                                                                                                                     
[build]   Abandoned: 3 packages were abandoned.                                                                                                                                              
[build]   Failed:    1 packages failed.                                                                                                                                                      
[build] Runtime: 17 minutes and 11.5 seconds total.                                                                                                                                          
[build] Note: Workspace packages have changed, please re-source setup files to use them.
```

There are a bunch of errors/warnings that show up when it fails, but here are a few of them:

```
In file included from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/pcl_nodelet.h:53:0,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/surface/convex_hull.h:41,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/surface/convex_hull.cpp:40:
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h: In instantiation of ‘boost::shared_ptr<pcl::PointCloud<PointT> > ros::DefaultMessageCreator<pcl::PointCloud<PointT> >::operator()() [with T = pcl::PointXYZ]’:
/usr/include/boost/function/function_template.hpp:138:22:   required from ‘static R boost::detail::function::function_obj_invoker0<FunctionObj, R>::invoke(boost::detail::function::function_buffer&) [with FunctionObj = ros::DefaultMessageCreator<pcl::PointCloud<pcl::PointXYZ> >; R = boost::shared_ptr<pcl::PointCloud<pcl::PointXYZ> >]’
/usr/include/boost/function/function_template.hpp:940:38:   required from ‘void boost::function0<R>::assign_to(Functor) [with Functor = ros::DefaultMessageCreator<pcl::PointCloud<pcl::PointXYZ> >; R = boost::shared_ptr<pcl::PointCloud<pcl::PointXYZ> >]’
/usr/include/boost/function/function_template.hpp:728:7:   required from ‘boost::function0<R>::function0(Functor, typename boost::enable_if_c<(boost::type_traits::ice_not<(boost::is_integral<Functor>::value)>::value), int>::type) [with Functor = ros::DefaultMessageCreator<pcl::PointCloud<pcl::PointXYZ> >; R = boost::shared_ptr<pcl::PointCloud<pcl::PointXYZ> >; typename boost::enable_if_c<(boost::type_traits::ice_not<(boost::is_integral<Functor>::value)>::value), int>::type = int]’
/usr/include/boost/function/function_template.hpp:1077:16:   required from ‘boost::function<R()>::function(Functor, typename boost::enable_if_c<(boost::type_traits::ice_not<(boost::is_integral<Functor>::value)>::value), int>::type) [with Functor = ros::DefaultMessageCreator<pcl::PointCloud<pcl::PointXYZ> >; R = boost::shared_ptr<pcl::PointCloud<pcl::PointXYZ> >; typename boost::enable_if_c<(boost::type_traits::ice_not<(boost::is_integral<Functor>::value)>::value), int>::type = int]’
/opt/ros/kinetic/include/message_filters/subscriber.h:146:7:   required from ‘void message_filters::Subscriber<M>::subscribe(ros::NodeHandle&, const string&, uint32_t, const ros::TransportHints&, ros::CallbackQueueInterface*) [with M = pcl::PointCloud<pcl::PointXYZ>; std::__cxx11::string = std::__cxx11::basic_string<char>; uint32_t = unsigned int]’
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/surface/convex_hull.cpp:71:51:   required from here
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:79:37: error: no match for ‘operator=’ (operand types are ‘std::shared_ptr<std::vector<pcl::detail::FieldMapping> >’ and ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >’)
       pcl::detail::getMapping(*msg) = mapping_;
                                     ^
```

```
/usr/include/c++/5/bits/shared_ptr.h:308:2: note:   template argument deduction/substitution failed:
In file included from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/pcl_nodelet.h:53:0,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/features/feature.h:45,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/features/boundary.h:44,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/features/boundary.cpp:39:
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:79:37: note:   ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >’ is not derived from ‘std::unique_ptr<_Tp, _Dp>’
       pcl::detail::getMapping(*msg) = mapping_;
                                     ^
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h: In instantiation of ‘static void ros::serialization::Serializer<pcl::PointCloud<PointT> >::read(Stream&, pcl::PointCloud<PointT>&) [with Stream = ros::serialization::IStream; T = pcl::PointXYZ]’:
/opt/ros/kinetic/include/ros/serialization.h:163:22:   required from ‘void ros::serialization::deserialize(Stream&, T&) [with T = pcl::PointCloud<pcl::PointXYZ>; Stream = ros::serialization::IStream]’
/opt/ros/kinetic/include/ros/subscription_callback_helper.h:136:21:   required from ‘ros::VoidConstPtr ros::SubscriptionCallbackHelperT<P, Enabled>::deserialize(const ros::SubscriptionCallbackHelperDeserializeParams&) [with P = const ros::MessageEvent<const pcl::PointCloud<pcl::PointXYZ> >&; Enabled = void; ros::VoidConstPtr = boost::shared_ptr<const void>]’
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/features/boundary.cpp:74:1:   required from here
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:209:85: error: invalid initialization of reference of type ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >&’ from expression of type ‘std::shared_ptr<std::vector<pcl::detail::FieldMapping> >’
         boost::shared_ptr<pcl::MsgFieldMap>& mapping_ptr = pcl::detail::getMapping(m);
                                                                                     ^
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h: In instantiation of ‘static void ros::serialization::Serializer<pcl::PointCloud<PointT> >::read(Stream&, pcl::PointCloud<PointT>&) [with Stream = ros::serialization::IStream; T = pcl::Normal]’:
/opt/ros/kinetic/include/ros/serialization.h:163:22:   required from ‘void ros::serialization::deserialize(Stream&, T&) [with T = pcl::PointCloud<pcl::Normal>; Stream = ros::serialization::IStream]’
/opt/ros/kinetic/include/ros/subscription_callback_helper.h:136:21:   required from ‘ros::VoidConstPtr ros::SubscriptionCallbackHelperT<P, Enabled>::deserialize(const ros::SubscriptionCallbackHelperDeserializeParams&) [with P = const ros::MessageEvent<const pcl::PointCloud<pcl::Normal> >&; Enabled = void; ros::VoidConstPtr = boost::shared_ptr<const void>]’
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/features/boundary.cpp:74:1:   required from here
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:209:85: error: invalid initialization of reference of type ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >&’ from expression of type ‘std::shared_ptr<std::vector<pcl::detail::FieldMapping> >’
```

```
/usr/include/c++/5/bits/shared_ptr.h:308:2: note:   template argument deduction/substitution failed:
In file included from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/pcl_nodelet.h:53:0,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/features/feature.h:45,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/features/moment_invariants.h:42,
                 from /home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/features/moment_invariants.cpp:39:
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:79:37: note:   ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >’ is not derived from ‘std::unique_ptr<_Tp, _Dp>’
       pcl::detail::getMapping(*msg) = mapping_;
                                     ^
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h: In instantiation of ‘static void ros::serialization::Serializer<pcl::PointCloud<PointT> >::read(Stream&, pcl::PointCloud<PointT>&) [with Stream = ros::serialization::IStream; T = pcl::PointXYZ]’:
/opt/ros/kinetic/include/ros/serialization.h:163:22:   required from ‘void ros::serialization::deserialize(Stream&, T&) [with T = pcl::PointCloud<pcl::PointXYZ>; Stream = ros::serialization::IStream]’
/opt/ros/kinetic/include/ros/subscription_callback_helper.h:136:21:   required from ‘ros::VoidConstPtr ros::SubscriptionCallbackHelperT<P, Enabled>::deserialize(const ros::SubscriptionCallbackHelperDeserializeParams&) [with P = const ros::MessageEvent<const pcl::PointCloud<pcl::PointXYZ> >&; Enabled = void; ros::VoidConstPtr = boost::shared_ptr<const void>]’
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/features/moment_invariants.cpp:73:1:   required from here
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:209:85: error: invalid initialization of reference of type ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >&’ from expression of type ‘std::shared_ptr<std::vector<pcl::detail::FieldMapping> >’
         boost::shared_ptr<pcl::MsgFieldMap>& mapping_ptr = pcl::detail::getMapping(m);
                                                                                     ^
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h: In instantiation of ‘static void ros::serialization::Serializer<pcl::PointCloud<PointT> >::read(Stream&, pcl::PointCloud<PointT>&) [with Stream = ros::serialization::IStream; T = pcl::Normal]’:
/opt/ros/kinetic/include/ros/serialization.h:163:22:   required from ‘void ros::serialization::deserialize(Stream&, T&) [with T = pcl::PointCloud<pcl::Normal>; Stream = ros::serialization::IStream]’
/opt/ros/kinetic/include/ros/subscription_callback_helper.h:136:21:   required from ‘ros::VoidConstPtr ros::SubscriptionCallbackHelperT<P, Enabled>::deserialize(const ros::SubscriptionCallbackHelperDeserializeParams&) [with P = const ros::MessageEvent<const pcl::PointCloud<pcl::Normal> >&; Enabled = void; ros::VoidConstPtr = boost::shared_ptr<const void>]’
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/src/pcl_ros/features/moment_invariants.cpp:73:1:   required from here
/home/mfe/code/voxbox_ws/src/pcl_catkin/perception_pcl/pcl_ros/include/pcl_ros/point_cloud.h:209:85: error: invalid initialization of reference of type ‘boost::shared_ptr<std::vector<pcl::detail::FieldMapping> >&’ from expression of type ‘std::shared_ptr<std::vector<pcl::detail::FieldMapping> >’
make[2]: *** [CMakeFiles/pcl_ros_features.dir/src/pcl_ros/features/shot_omp.cpp.o] Error 1
make[2]: *** [CMakeFiles/pcl_ros_features.dir/src/pcl_ros/features/moment_invariants.cpp.o] Error 1
make[1]: *** [CMakeFiles/pcl_ros_features.dir/all] Error 2
make: *** [all] Error 2
```

Any suggestions for what is causing this? Thank you!

**Comment by margaritaG** [comment]
Hi Michael,

Thank you for pointing out this issue. The problem is caused by some recent changes introduced in PCL, which break compilation of _pcl_ros_ (see https://github.com/ethz-asl/pcl_catkin/pull/20).

I have now set the specific version of PCL to be checked out to 1.9.1 (not the latest master commit) and the issue should be now resolved.

**Comment by mfe7** [comment]
Thanks @margaritaG , unfortunately I just pulled the latest commit and saw the version of pcl get adjusted, re-ran `wstool update`, `catkin clean` and `catkin build mask_rcnn_ros depth_segmentation gsm_node` but got the same errors.

**Comment by margaritaG** [comment]
@mfe7 It seems that `wstool update` is not enough to actually checkout the newly specified version (branch) of `pcl_catkin`.

Try first `wstool merge -t . voxblox-plusplus/voxblox-plusplus_https.rosinstall`, it should ask something like this:
```
$ wstool merge -t . voxblox-plusplus/voxblox-plusplus_https.rosinstall

     Change details of element (Use --merge-keep or --merge-replace to change):
  pcl_catkin

Continue: (y)es, (n)o, (v)erbosity, (a)dvanced options: 

```

After this run `wstool update`.

Else, you can just manually checkout the branch `fix/pcl-1.9.1` in the `pcl_catkin` folder.

Let me know if any of these solve the issue.

**Comment by mfe7** [comment]
@margaritaG Yep the `wstool merge` fixed it. Thank you!
